docs_src/design/adr/security/0019-spiffe.md

```diff
@@ -0,0 +1,338 @@
+  It is simply easier to run without security enabled in Kubernetes.
+
+* The current approach cannot support distributed services
+  without an underlying distributed file system to distribute tokens,
+  such as GlusterFS, running across the participating nodes.
+
+## Decision
+
+EdgeX will create a new service, `security-spiffe-token-provider`.
+This service will be a mutual-auth TLS service that exchanges
+a [SPIFFE](https://spiffe.io/) X.509 SVID for a secret store token.
+
+The SPIFFE token provider will take theee parameters:
+
+1. An X.509 SVID used in mutual-auth TLS for the token provider
+   and the service to cross-authenticate.
+   
+2. The reqested service key.  If blank, the service key will
+   default to the service name encoded in the SVID.
+   If the service name follows the pattern `device-(name)`,
+   then the service key must follow the format
+   `device-(name)` or `device-name-*`.
+   If the service name is `app-service-configurable`,
+   then the service key must follow the format
+   `app-*`.
+
+3. A list of "known secret" identifiers that will allow
+   new services to request database passwords or
+   other "known secrets" to be seeded into their service's
+   partition in the secret store.
+
+The `go-mod-secrets` module will be modified to enable a new mode
+whereby a secret store token is obtained by:
+
+1. Obtaining an X.509 SVID by contacting a local SPIFFE agent's
+   workload API on a local Unix domain socket.
+2. Connecting to the `security-spiffe-token-provider` service
+   and using the X.509 SVID and request a secret store token.
+
+The SPIFFE authentication mode will be an opt-in feature.
+
+This feature is estimated to be a "large" or "extra large"
+effort that could be implemented in a single release cycle.
+
+### Technical Architecture
+
+![SPIFFE Architecture and Workflow](0019-spiffe-architecture.jpg)
+
+The work flow is as follows:
+
+1. The SPIFFE server is started.
+1. The server creates an initial set of keys.
+1. The trust bundle (certificate authority) data is exported from the SPIFFE server
+   and workload entries are loaded into the SPIFFE server database.
+1. A join token for the SPIFFE agent is created.
+1. The SPIFFE agent is started with the join token to add it to the cluster.
+1. Vault is started and `security-secret-store-setup`
+   initializes it and creates an admin token for `security-spiffe-token-provider` to use.
+1. The `security-spiffe-token-provider` service is started.
+   It obtains an SVID from the SIFFE agent and uses it as a TLS server certificate.
+1. An EdgeX microservice starts and obtains another SVID from the SPIFFE agent
+   and uses it as a TLS client certificate to contact the
+   `security-spiffe-token-provider` service.
+   The EdgeX microservice uses the trust bundle as a server CA
+   to verify the TLS certificate of the remote service.
+1. `security-spiffe-token-provider` verifies the SVID using the trust bundle as client CA
```

Comment:
And just to be clear, this provider is only used to get the initial token, renewal is handled directly via vault, correct?

Can a service be enabled for some period of time, disabled, and then re-enabled again?